pkg/cliplugins/workspace/plugin/kubeconfig.go

```diff
@@ -673,3 +675,109 @@ func newKCPClusterClient(clientConfig clientcmd.ClientConfig) (kcpclient.Cluster
+type ShowWorkspaceTreeOptions struct {
+	*base.Options
+
+	kcpClusterClient kcpclient.ClusterInterface
+}
+
+// NewShowWorkspaceTreeOptions returns a new ShowWorkspaceTreeOptions.
+func NewShowWorkspaceTreeOptions(streams genericclioptions.IOStreams) *ShowWorkspaceTreeOptions {
+	return &ShowWorkspaceTreeOptions{
+		Options: base.NewOptions(streams),
+	}
+}
+
+// BindFlags binds fields to cmd's flagset.
+func (o *ShowWorkspaceTreeOptions) BindFlags(cmd *cobra.Command) {}
+
+// Complete ensures all dynamically populated fields are initialized.
+func (o *ShowWorkspaceTreeOptions) Complete() error {
+	if err := o.Options.Complete(); err != nil {
+		return err
+	}
+
+	kcpClusterClient, err := newKCPClusterClient(o.ClientConfig)
+	if err != nil {
+		return err
+	}
+	o.kcpClusterClient = kcpClusterClient
+
+	return nil
+}
+
+// Run outputs the current workspace.
+func (o *ShowWorkspaceTreeOptions) Run(ctx context.Context) error {
+	config, err := o.ClientConfig.ClientConfig()
+	if err != nil {
+		return err
+	}
+	_, currentClusterName, err := pluginhelpers.ParseClusterURL(config.Host)
+	if err != nil {
+		return fmt.Errorf("current URL %q does not point to cluster workspace", config.Host)
+	}
+
+	tree := treeprint.New()
+	b := tree.AddBranch("workspaces")
+	var result *v1beta1.WorkspaceList
+
+	if err := wait.PollImmediate(time.Millisecond*100, time.Second*5, func() (bool, error) {
+		if result, err = o.kcpClusterClient.Cluster(currentClusterName).TenancyV1beta1().Workspaces().List(ctx, metav1.ListOptions{}); err != nil {
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		}
+		return true, nil
+	}); err != nil {
+		return err
+	}
+
+	for _, workspace := range result.Items {
+		_, currentClusterName, err := pluginhelpers.ParseClusterURL(workspace.Status.URL)
+		if err != nil {
+			return fmt.Errorf("current URL %q does not point to cluster workspace", config.Host)
+		}
+		err = populateBranch(ctx, b, currentClusterName, o.kcpClusterClient)
+		if err != nil {
+			return err
+		}
+	}
+
+	fmt.Println(tree.String())
+	return nil
+}
+
+func populateBranch(ctx context.Context, tree treeprint.Tree, name logicalcluster.Name, kcpClusterClient kcpclient.ClusterInterface) error {
+	var result *v1beta1.WorkspaceList
+	var err error
+	if err := wait.PollImmediate(time.Millisecond*100, time.Second*5, func() (bool, error) {
+		if result, err = kcpClusterClient.Cluster(name).TenancyV1beta1().Workspaces().List(ctx, metav1.ListOptions{}); err != nil {
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		}
+		return true, nil
+	}); err != nil {
+		return err
+	}
+
+	b := tree.AddBranch(name.String())
+
+	for _, workspace := range result.Items {
+		_, currentClusterName, err := pluginhelpers.ParseClusterURL(workspace.Status.URL)
+		if err != nil {
+			return fmt.Errorf("current URL %q does not point to cluster workspace", workspace.Status.URL)
+		}
+		if !strings.Contains(workspace.Name, ":") {
```

Comment:
To my knowledge it's not possible to have `:` in a name, anyway. We shouldn't need to do this check.

Reply:
yap, was me trying to understand how to construct this better. Should be ok now